File: src/main/java/puzzles/year24/day5/Part2.java
package puzzles.year24.day5;

import util.Puzzle;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;

public class Part2 implements Puzzle<Integer>  {
    @Override
    public Integer getAnswer(String inputString) {
        String[] parts = inputString.split(System.lineSeparator() + System.lineSeparator());
        String[] ruleString = parts[0].split(System.lineSeparator());

        HashMap<Integer, List<Integer>> rules = new HashMap<>();

        for (String rule : ruleString) {
            String[] s = rule.split("\\|");
            rules.compute(Integer.parseInt(s[0]), (integer, integers) -> {
                if (integers == null) integers = new ArrayList<>();
                integers.add(Integer.parseInt(s[1]));
                return integers;
            });
        }

        String[] lines = parts[1].split(System.lineSeparator());

        int result = 0;
        for (String line : lines) {
            Set<Integer> seen = new HashSet<>();
            List<Integer> pages = new ArrayList<>(Arrays.stream(line.split(",")).map(Integer::parseInt).toList());
            boolean bad = false;
            for (int i = 0; i < pages.size(); i++) {
                int page = pages.get(i);
                seen.add(page);

                List<Integer> rule = rules.get(page);

                if (rule != null) {
                    Optional<Integer> seenValue = rule.stream().filter(seen::contains).findFirst();

                    if (seenValue.isPresent()) {
//                        System.out.println("Swapping " + seenValue.get() + " and " + pages.get(i));
                        swap(pages, seenValue.get(), i);
//                        pages.add(i + 1, seenValue.get());
//                        pages.remove(seenValue.get());
//                        pages.remove(i);
//                        pages.add(pages.indexOf(seenValue.get()), page);



                        i = -1;
                        seen = new HashSet<>();
                        bad = true;
                    }
                }
            }

            if (bad) {
                result += pages.get(pages.size() / 2);
//                System.out.println(line);
//                System.out.println(String.join(",", pages.stream().map(String::valueOf).toList()));
            }

        }
        return result;
    }

    private void swap(List<Integer> list, Integer first, int secondIndex) {
        int firstIndex = list.indexOf(first);

        list.set(firstIndex, list.get(secondIndex));
        list.set(secondIndex, first);
    }
}

File: src/main/java/puzzles/year24/day23/Part2.kt
package puzzles.year24.day23

import util.Puzzle

class Part2 : Puzzle<String?> {
    override fun getAnswer(inputString: String): String {
        val lines = inputString.lines()
        val graph = mutableMapOf<String, MutableSet<String>>()

        for (line in lines) {
            val parts = line.split("-")
            graph.compute(parts[0]) {_, v -> add(v ?: mutableSetOf(), parts[1])}
            graph.compute(parts[1]) {_, v -> add(v ?: mutableSetOf(), parts[0])}
        }

        for (computer in graph.keys) {
            val connections = graph[computer]!!
            val subsets = subsets(connections)

            for (subset in subsets) {
                subset.add(computer)

                if (isAllConnected(graph, subset)) {
                    return subset.sorted().joinToString(",")
                }
            }
        }


        return ""
    }

    fun isAllConnected(graph: MutableMap<String, MutableSet<String>>, set: Set<String>): Boolean {
        for (computer in set) {
            for (computer2 in set) {
                if (computer != computer2 && !graph[computer].orEmpty().contains(computer2)) {
                    return false
                }
            }
        }
        return true
    }

    fun subsets(set: Set<String>): MutableSet<MutableSet<String>> {
        val result = mutableSetOf<MutableSet<String>>()
        for (x in set) {
            val newSet = HashSet(set)
            newSet.remove(x)
            result.add(newSet)
        }

        return result
    }

    fun <T>add(set: MutableSet<T>, value: T): MutableSet<T> {
        set.add(value)
        return set
    }
}
